pymacaroons/serializers/json_serializer.py

```diff
@@ -1,39 +1,214 @@
+                verification_key_id=_read_json_binary_field('v', c),
+                location=_read_json_binary_field('l', c)
+            )
+            caveats.append(caveat)
+        return Macaroon(
+            location=_read_json_binary_field('l', deserialized),
+            identifier=_read_json_binary_field('i', deserialized),
+            caveats=caveats,
+            signature=_read_json_binary_field('s', deserialized),
+            version=MACAROON_V2
         )
+
+
+def _caveat_v1_to_dict(c):
+    ''' Return a caveat as a dictionary for export as the JSON
+    macaroon v1 format.
+    '''
+    serialized = {}
+    if len(c.caveat_id) > 0:
+        serialized['cid'] = utils.convert_to_string(c.caveat_id)
+    if c.verification_key_id:
+        serialized['vid'] = base64.urlsafe_b64encode(
+            c.verification_key_id).decode('utf-8')
+    if c.location:
+        serialized['cl'] = c.location
+    return serialized
+
+
+def _caveat_v2_to_dict(c):
+    ''' Return a caveat as a dictionary for export as the JSON
+    macaroon v2 format.
+    '''
+    serialized = {}
+    if len(c.caveat_id) > 0:
+        _add_json_binary_field(c.caveat_id, 'i', serialized)
+    if c.verification_key_id:
+        _add_json_binary_field(c.verification_key_id, 'v', serialized)
+    if c.location:
+        serialized['l'] = c.location
+    return serialized
+
+
+def _add_json_binary_field(b, field, serialized):
+    ''' Add value of b into one of the fields depending on its value.
+    '''
+    try:
+        val = b.decode("utf-8")
+        serialized[field] = val
+    except UnicodeDecodeError:
+        val = _raw_urlsafe_b64encode(b).decode('utf-8')
+        serialized[field + '64'] = val
+
+
+def _read_json_binary_field(field, deserialized):
+    ''' Read the value of a JSON field that may be string or base64-encoded.
+    '''
+    val = deserialized.get(field)
+    if val is not None:
+        return val
+    val = deserialized.get(field + '64')
+    if val is None:
+        return None
+    return _raw_urlsafe_b64decode(val)
+
+
+def _add_base64_padding(b):
+    '''Add padding to base64 encoded bytes.
+
+    Padding can be removed when sending the messages.
+
+    @param b bytes to be padded.
+    @return a padded bytes.
+    '''
+    return b + b'=' * (-len(b) % 4)
+
+
+def _remove_base64_padding(b):
+    '''Remove padding from base64 encoded bytes.
+
+
+    @param bytes b to be padded.
+    @return a padded bytes.
+    '''
+
+    return b.rstrip(b'=')
+
+
+def _raw_urlsafe_b64decode(s):
+    '''Base64 decode with added padding and conversion to bytes.
+
+    @param s string decode
+    @return bytes decoded
+    '''
+    return base64.urlsafe_b64decode(_add_base64_padding(
+        s.encode('utf-8')))
+
+
+def _raw_urlsafe_b64encode(b):
+    '''Base64 encode with padding removed.
+
+    @param s string decode
+    @return bytes decoded
+    '''
+    return _remove_base64_padding(base64.urlsafe_b64encode(b))
```

Comment:
Honestly, I'd just lose _remove_base64_padding and just do;

     return base64.urlsafe_b64encode(b).rstrip(b'=')